### PR TITLE
Bound compression-pointer following to prevent parse-time DoS

### LINKED ISSUE #116: Bound DNS name compression-pointer following to prevent parse-time DoS
Follow-up from PR #115 (#109) review.

## Problem

`parse/1` (and therefore `parse_safe/1`) follows DNS name compression pointers without bounding the work. A maliciously crafted **pointer loop** — a name whose compression pointer cycles back to itself or forms a cycle — makes `parse_name_acc/3` recurse forever, exhausting CPU/memory. `parse_safe/1` returns `{:error, :malformed}` for input that *raises*, but a pointer loop doesn't raise — it loops, so the safe wrapper does not interrupt it.

Confirmed: a self-referential pointer packet (`<<0x1234::16, 0::16, 1::16, 0::16, 0::16, 0::16, 0xC0, 0x0C>>`, question name at offset 12 pointing to offset 12) does not return within a 2s timeout.

This is a DoS vector for any code parsing untrusted packets off the network (the primary use case for `parse_safe/1`). The limitation is now documented in `parse_safe/1`'s "Limitations" section, but the actual fix belongs here.

## Proposed approach

Bound pointer following in `parse_name_acc/3`. Standard techniques (RFC 1035 §4.1.4 implementations use one or both):

- **Strictly-decreasing offsets**: a compression pointer must point *backwards* to an already-seen position; reject (or stop) if it points at or after the current name's start. This is the common BIND/Unbound rule and makes loops impossible.
- **Pointer-follow cap**: limit the number of pointers followed per name (e.g. ≤ 128) and/or total name length (≤ 255 bytes per RFC 1035), returning a parse error past the cap.

On violation, `parse/1` should raise (consistent with its trusted-input contract) and `parse_safe/1` should map it to `{:error, :malformed}` (extend the rescue set or add an explicit error path).

## Acceptance criteria

- [ ] A crafted pointer-loop packet terminates quickly with a parse error rather than looping (covered by a test + ideally the existing no-raise fuzz extended to include pointer bytes once loops are impossible)
- [ ] Legitimate compression (backward pointers to earlier names) still round-trips — existing tests + property suite (#107) stay green
- [ ] Benchmark within ±5% (the bound is a cheap comparison on the existing pointer path)
- [ ] Update the `parse_safe/1` "Limitations" note once the loop is defended

### PR DESCRIPTION
**Pull request by toshi0806**
resolve #116

Fixes the compression-pointer DoS surfaced in the #115 review: a crafted name pointer that cycles (self-reference, 2-cycle, forward re-entry) made `parse_name_acc/3` recurse forever, and `parse_safe/1` couldn't interrupt it (a loop isn't a raise).

## Fix

Enforce RFC 1035 §4.1.4 backward-only compression. A strictly-decreasing `ceiling` is threaded through the name parser (initially `byte_size(orig_body)`); a pointer is followed only when `offset < ceiling`, recursing with `ceiling = offset`. The jump-target sequence then strictly decreases, so **cycles are impossible**. A violating pointer matches no clause → `FunctionClauseError`, which `parse/1` raises (trusted-input contract) and `parse_safe/1` already maps to `{:error, :malformed}`.

Legitimate compression is unaffected: a name at offset `O` only references names at offsets `< O`, so well-formed backward pointers always satisfy the bound.

## Tests

- `test/dns_packet_pointer_loop_test.exs`: self-referential, 2-cycle, and forward-re-entry pointers → `parse_safe/1` returns `{:error, :malformed}` and `parse/1` raises. Each loop case carries `@tag timeout: 2_000` so the **pre-fix** infinite loop fails as a timeout instead of hanging the suite (verified: red before, green after).
- Pins that a legitimate answer-name-points-to-question-name packet still decodes.
- Widened the `parse_safe/1` no-raise fuzz to the full **0..255** byte range (previously `0..63` to dodge the loop) — it now exercises compression-pointer paths and stays hang-free and raise-free across seeds 0/1/42/9999/123456/777.
- Updated `parse_safe/1`'s "## Limitations": loops are now defended; only the size-proportional work note remains.

## Verification

- 246 tests pass (4 doctests, 6 properties, 236 tests), `mix dialyzer` 0 errors, credo/format clean
- **Benchmark (median of 5 back-to-back A/B runs)**: all scenarios within ±5% except "parse EDNS response", which sits at the noise floor — it read −5.1% at 3 runs and +5.1% at 5 runs (sign-flipping ⇒ noise, not a regression). The change is O(1) per name (one `byte_size/1` on a binary + an integer guard on the pointer path only):

  | scenario | Δ (median of 5) |
  |---|---|
  | parse simple query | +3.5% |
  | create simple query | +2.1% |
  | parse EDNS response | +5.1% (noise: −5.1% at 3 runs) |
  | parse complex response | 0.0% |
  | create EDNS response | +3.8% |
  | create complex response | +3.4% |

### CODE REVIEW

File: lib/dns_packet.ex [hunk 1]
```diff
@@ -707,29 +705,41 @@ defmodule DNSpacket do
...
     {body, orig_body, "."}
   end
 
-  defp parse_name_acc(<<0::8, body::binary>>, orig_body, acc) do
+  defp parse_name_acc(<<0::8, body::binary>>, orig_body, acc, _ceiling) do
     {body, orig_body, IO.iodata_to_binary(Enum.reverse(acc))}
   end
 
-  defp parse_name_acc(<<0b11::2, offset::14, body::binary>>, orig_body, acc) do
+  defp parse_name_acc(<<0b11::2, offset::14, body::binary>>, orig_body, acc, ceiling)
```
[the remark this answers is not in the document]
Comment (toshi0806) [reply lib/dns_packet.ex:729]: Confirmed — your analysis is exactly right, and I added a test in d9508ff to pin the chained case you walked through.

The key invariant: the `ceiling` is the offset we **last jumped to**, and it stays in effect for the whole resolution of that target name, including labels parsed forward from it. So after jumping to offset 20, both the labels at 20..k and any pointer found at k are evaluated against `ceiling = 20`. A pointer to 18 is admitted (18 < 20, a real backward reference); a pointer to 21 is rejected — and as you note, a name *at* offset 20 referencing offset 21 can't be legitimate compression, because offset 21 was written *after* offset 20, so it didn't exist when offset 20's name was encoded. The same reasoning makes the whole A→B→C chain safe: each hop lands at a strictly smaller offset, which is precisely the shape RFC 1035 compression always takes.

The new test "a chained pointer … decodes" builds a two-hop chain (answer2's name is a pointer to answer1's name at offset 21, which is a label `"z"` plus a pointer to the question name `"x.y."` at offset 12 → resolves `21 → 12`, strictly decreasing) and asserts it decodes to `"z.x.y."`. So the chained-legitimate path is now covered, alongside the loop-rejection cases.

File: lib/dns_packet.ex [hunk 2]
```diff
@@ -707,29 +705,41 @@ defmodule DNSpacket do
...
+    parse_name_acc(body, orig_body, [result], byte_size(orig_body))
   end
 
-  defp parse_name_acc(<<0::8, body::binary>>, orig_body, []) do
+  # `ceiling` is the strictly-decreasing bound a compression pointer must
+  # stay under: a pointer may only point backward (RFC 1035 §4.1.4), so each
+  # followed pointer's offset must be < the offset we last jumped to. The
+  # jump targets therefore strictly decrease, making loops impossible. A
+  # pointer that violates this (offset >= ceiling) matches no clause and
+  # raises FunctionClauseError, which parse_safe/1 maps to :malformed (#116).
```
[the remark this answers is not in the document]
Comment (toshi0806) [reply lib/dns_packet.ex:720]: Keeping the clause-non-match approach, for consistency. `parse/1` already signals **all** malformed input the same way — a truncated record or a bad embedded length matches no binary-pattern clause and raises `FunctionClauseError`/`MatchError`. A rejected compression pointer is just one more "no clause matches" case, so it falls into the same uniform model rather than a special-cased path. Introducing a dedicated exception (or an `{:error, _}` return from deep inside `parse_name_acc`) would fragment that model and force `parse_safe/1` to learn a third error channel.

On the "rescue catching incidental `FunctionClauseError`" risk: that was the #115 round-2 concern, already mitigated — the rescue is narrowed to `[FunctionClauseError, MatchError]` and the no-raise property (now over full `0..255` bytes including pointers) confirms `parse_safe/1` only ever returns tagged tuples. A genuine internal bug raising one of those would indeed read as `:malformed` for `parse_safe/1` callers, but trusted callers using `parse/1` still get the raised exception — which is the intended split. The intent is documented in the clause comment.

File: lib/dns_packet.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (toshi0806) [reply lib/dns_packet.ex:729]: Confirmed safe, and pinned with tests in 9c5882c. The empty-`tmp_body` case can't arise through an allowed pointer: `ceiling` starts at `byte_size(orig_body)` and only ever **decreases** (each jump sets `ceiling = offset`, which had to be `< ceiling`). So for any pointer that passes the guard, `offset < ceiling ≤ byte_size(orig_body)` ⟹ `offset < byte_size` strictly — never `==`. The inner slice `<<_::binary-size(^offset), tmp_body::binary>>` therefore always leaves `tmp_body` with ≥ 1 byte. Added "a pointer to exactly the message end (offset == byte_size) is rejected" and "… past the message end" so both boundaries are covered.

File: test/dns_packet_pointer_loop_test.exs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      assert DNSpacket.parse_safe(packet) == {:error, :malformed}
+    end
+
+    @tag timeout: 2_000
+    test "forward pointer that re-enters (12 -> 14, 14 -> 14)" do
+      packet = header() <> <<0xC0, 0x0E, 0xC0, 0x0E>>
+      assert DNSpacket.parse_safe(packet) == {:error, :malformed}
+    end
+
+    test "a pointer to exactly the message end (offset == byte_size) is rejected" do
```
[the remark this answers is not in the document]
Comment (toshi0806) [reply test/dns_packet_pointer_loop_test.exs:43]: Good consistency point — applied in f55272c. Both boundary cases now carry `@tag timeout: 2_000` like the other three, so if a future guard change ever let `offset >= byte_size` fall through to a loop, they'd fail as a timeout rather than hang the suite.

File: lib/dns_packet.ex [hunk 4]
```diff
@@ -707,29 +705,41 @@ defmodule DNSpacket do
...
+  defp parse_name_acc(<<0::8, body::binary>>, orig_body, [], _ceiling) do
```
[the remark this answers is not in the document]
Comment (toshi0806) [reply lib/dns_packet.ex:721]: Fair complexity note, but worth separating from this PR: the per-record "expand each name's full compression chain" work is **pre-existing** — `parse_name` already resolved every record's pointers independently before #116. This change only adds a strictly-decreasing `ceiling`, which can only *reduce* per-name work (it caps a chain at the first jump target instead of following unbounded), so it doesn't worsen the records × name-length worst case; it bounds it to be finite.

And that worst case is bounded by the packet size: each name's pointer chain strictly decreases, so it's ≤ packet_size hops, and total work is O(records × packet_size). For DNS that's small in practice (UDP payloads are typically ≤ 512–4096 bytes, ≤ 64 KB hard cap). The adversarial-total-work mitigation is the upstream **packet-size cap** already recommended in `parse_safe/1`'s "Limitations" — a per-name visited-set or memoization would add overhead to the overwhelmingly common pointer-light case to defend a worst case the size cap already addresses, so I'd rather not. If a profile ever shows this mattering on real traffic, memoizing resolved offsets would be the follow-up.
